### ISSUE TITLE
Create a devtools client for testing the server

### ISSUE THREAD

**Issue by jdm** [servo/servo#5971]
The devtools server lives in components/devtools and is used for external introspection of web content in Servo by the high-quality developer tools that live in Firefox. All testing is performed through Firefox's developer tools right now, but it would be a great benefit to create a harness that could run simple testcases automatically and check expected output for a given input.

We should write a simple client that connects to a server (ie. Servo running with --devtools) and takes input through stdin to send commands to servo. We will likely want to extract the various JSON-based message types from `components/devtools` into a separate shared `devtools_msg` crate for maximal code reuse. This could be a great way to play with a small independent project in Rust while benefiting Servo!

**Comment by akosel** [comment]
I will work on this.

**Comment by jdm** [comment]
@akosel Fantastic! Just shout if you have any questions :)

**Comment by akosel** [comment]
I have started looking into this. I noticed the comment in https://github.com/servo/servo/issues/5970, and I assume I can follow a similar structure for this? 

That said, there is one issue that I am currently working around. When I run `./mach run http://google.com --devtools` I get: 

```
stack backtrace:

   1:        0x11060620f - sys::backtrace::write::h6990067d88907b48zVr
   2:        0x110609be4 - panicking::on_panic::he976cdbadffd9e1bEVv
   3:        0x1105f56c5 - rt::unwind::begin_unwind_inner::ha519446d83876aa8nDv
   4:        0x1105f5efc - rt::unwind::begin_unwind_fmt::h15c2d4adf8e8ecdatCv
   5:        0x11060949c - rust_begin_unwind
   6:        0x11062e945 - panicking::panic_fmt::haa553fbfab6eff2cNKy
   7:        0x10e80064f - option::Option<T>::expect::h2190493558643449847
   8:        0x10e800528 - collections::hash::map::HashMap<K, V, S>.Index<&'a Q>::index::h11570370938463914438
   9:        0x10e7fad25 - run_server::find_console_actor::h36bbabe37c10d21agJd
  10:        0x10e7edd07 - run_server::handle_network_event::h3503e9b8696ae0364Jd
  11:        0x10e7e0b99 - run_server::h865311e4cea3d37e6yd
  12:        0x10e7dfa33 - start_server::closure.13443
  13:        0x10e7df8dc - task::spawn_named::closure.13437
  14:        0x10e7df805 - boxed::F.FnBox<A>::call_box::h2076594675215285636
  15:        0x10e7aa51b - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h7459643712465505719
  16:        0x10e7a9b7b - thread::Builder::spawn_inner::closure.12797
  17:        0x10e7a9afe - rt::unwind::try::try_fn::__rust_abi::h742564627405228215
  18:        0x10e7a9a99 - rt::unwind::try::try_fn::h742564627405228215
  19:        0x11060b8d8 - rust_try_inner
  20:        0x11060b8c5 - rust_try
  21:        0x10e7a8f5a - rt::unwind::try::h12091752746280580590
  22:        0x10e7a8d04 - thread::Builder::spawn_inner::closure.12690
  23:        0x10e7aa85d - boxed::F.FnBox<A>::call_box::h15314979547732952471
  24:        0x11060856d - sys::thread::Thread::new::thread_start::h5fcf8fe12ec3c274mYu
  25:     0x7fff84fd3267 - _pthread_body
  26:     0x7fff84fd31e4 - _pthread_start
```

This is my first dive into servo, so some of this may be expected based on the current state of the project. So the problem is that HashMap[“any_value”] will panic! if the lookup fails. The bigger problem seems to be that when `find_console_actor` is initially called, `actor_pipelines` is empty, so any lookup will fail. I am getting around this by doing the lookup with the .get() method (to avoid the panic!), but that seems like it could easily lead to later failures. 

I am happy to file this as a bug if it is, in fact, a bug (and not just the result of me doing something incorrectly, which is possible!).

**Comment by jdm** [comment]
That is definitely a bug that should be filed. I wonder if this behaviour changed in the most recent Rust upgrade...

**Comment by jdm** [comment]
Oh, it's a regression from #5947 - we should avoid panicking if there's no console actor available!

**Comment by jdm** [comment]
#6540 should merge shortly. @akosel are you still intending to work on this?

**Comment by akosel** [comment]
Yes! I have a fork which has much of the preliminary work done. I will dedicate this weekend to finishing as much of this as possible.

**Comment by jdm** [comment]
Looking forward to it!

**Comment by akosel** [comment]
I have made some progress. I am at a point where I want to check in to make sure I am heading in the right direction. 

My first question is about extracting the JSON-based message types from `components/devtools`. I wasn’t exactly sure what the scope of that was, so for starters I grabbed various message types from the main lib.rs file as well as messages from the actors directory. Basically, I took everything that had the `#[derive(RustcEncodable)]` adornment. However, some of the messages took types specific to `devtools/components`, so I wasn’t 100% sure what to do there. For example, NetworkEventMsg struct requires `actors::network_event::EventActor` for one of its fields, which made it hard to extract cleanly. This may not be a big issue, but I wanted to touch base before doing anything more substantial.

The second question is more architectural. I currently set up a separate devtools_client crate. The crate is set up as a binary program. Beyond this, what should I add (if anything) to the `tests` directory? I was reading through the [Mozilla Remote Debugging Protocol Wiki](https://wiki.mozilla.org/Remote_Debugging_Protocol) and figured that would be a good place to start for writing test cases. I will need to create some new JSON-based messages that the client can send to servo. My best guess is that these messages should live in the tests directory, but I wanted to make sure. 

Also, as a community-best-practices question, what is the policy for PRs within Servo? Are they only created once the code is believed to be ready for merging, or is it acceptable to open a PR to make it easier for someone to review code in an intermediate state? I ask this because I am wondering if it is ok to open a PR to make it easier to review my code in its current state.

**Comment by jdm** [comment]
These are all great questions! With regards to the EventActor, that's a struct that you should be able to extract without any difficulties I believe. It's just a Plain Old Data struct for serialization.

My intuition is that the new messages could be added to the devtools_client crate. Additionally, instead of a binary, if the devtools_client crate is a library we can then do the following:
* create `tests/devtools`
* create a `harness.rs` which has methods to spawn an instance of Servo with a devtools server running, and methods to initiate a connection to the server
* every other Rust file in that directory can be a test that calls these methods, then sends messages (defined in `devtools_msg`) and inspects the responses for expected values

These can be integrated similar to the tests in the `test-unit` suite (http://mxr.mozilla.org/servo/source/python/servo/testing_commands.py#134), I think. We could even just add tests in `tests/unit/devtools_client`, I suppose, and run them using `./mach test-unit -p devtools_client`.

**Comment by jdm** [comment]
Oh and feel free to make PRs whenever you'd like. Just make your expectations clear up front for reviewers if you don't feel like the code is actually ready for integration.

**Comment by jdm** [comment]
There's exciting stuff going on here!

**Comment by akosel** [comment]
Yes, very exciting! Thanks again for the patience. This has been enjoyable to work on. If there becomes a point where this is mission critical, I can hand it off to someone who is more experienced. I would prefer to finish what I started if possible, but I also don't want to slow everything down! :)

As an update, I'm currently in a position where I can spin up a servo instance running devtools, connect to it with the devtools client, and send it an encodable message.

All that said, I do have three more questions. I am currently using `std::process::Command` to spawn a servo instance, which doesn't seem like a terrible idea, but it might be. Do you have any input on a better way to do this?

A second, somewhat related question is in regards to sending a signal to a servo instance that it should die (after running the tests). I have a feeling I can send a message over the port I connect to with the client to signal servo that it should shutdown. I could use some hints on where I would look to learn about that. 

A third, also somewhat related question/dilemma surrounds how I can signal to a client that servo is up, running, and listening for connections. My current hack is to put the thread that starts the client to sleep for a second, which is not a reliable way to handle this. I could also loop until a connection is established, but that would be a bad idea if for some reason servo fails to start the devtools server. I guess the ideal scenario would be subscribing to a channel used by the devtools server that let's me know when (or if) it is listening, but I don't believe this exists at the moment. Would it make sense to build a channel like this? Or does a channel already exist that I can access through some means that I am not seeing?

Thank you again for all of your help. This has been a great way to get better acquainted with Rust.

**Comment by jdm** [comment]
Sorry for the delay in answering these!
1) I don't really have an opinion about using Command; I'd say if it does what we need then it's fine.
2) There is currently no intentional way to shut down the server from the client. That being said, evaluating "window.close()" in the scope of the current pipeline would probably work. On the other hand, it probably makes more sense to just [kill](http://doc.servo.org/std/process/struct.Child.html#method.kill) the child process instead.
3) I'm not really bothered by trying to connect repeatedly until it works. We could probably just time out and fail the test after 30s or something (just check the difference between current and start time each time before we try to connect to the sever).

**Comment by jdm** [comment]
If we were launching the browser in a thread rather than using a process, we could add some sort of Sender to http://mxr.mozilla.org/servo/source/components/servo/lib.rs#79 and http://mxr.mozilla.org/servo/source/components/servo/lib.rs#105, but I'm not convinced that it's worth the extra effort.

**Comment by akosel** [comment]
Ok, this all sounds good. Thank you for the insights on shutting down the server. I am struggling a bit with calling child.kill from within a thread (because of ownership issues), while child.wait is blocking the main thread. At least I know this struggle is worth figuring out. I will also look at potentially using window.close(). Thanks!

**Comment by jdm** [comment]
Could we spawn a separate thread for the client, and when the client wants to shut down it sends a message to the original thread which is just waiting in a recv loop and presumably has ownership of the child process and can kill it?

**Comment by akosel** [comment]
Yes, that seems like the right way to do things. I appreciate the guidance!

**Comment by KiChjang** [comment]
Is this complete?

**Comment by jdm** [comment]
It is not complete. Certainly it would be useful to look at the commits on [@akosel's branch](https://github.com/akosel/servo) if anybody else chooses to pick this up.

**Comment by jdm** [comment]
Having thought further about it, I don't yet have a good idea of how to control the actions of the server.

**Comment by pkondzior** [comment]
@jdm what is left on this issue ?

**Comment by jdm** [comment]
@pkondzior: Everything. We need a client that can connect to the server and assert properties about the messages that it receives, and we also need a way to control the behaviour of Servo. We probably want to do that via the webdriver server.

**Comment by jdm** [comment]
I think this has been addressed by https://jpramosi.github.io/geckordp/ which we use in our unit tests for the devtools server (`./mach test-devtools`).
